### ISSUE TITLE
Console not working with NestedStacks

### ISSUE THREAD

**Issue by romcok** [sst/sst#1543]
Hi, due to API size and deployment issues, we have split API into nested stacks. I tried to run the console in the project and it got the message "Syncing metadata". I see in the inspector that this nested stack error is repeated:

Request URL: http://localhost:12557/proxy/https://cloudformation.eu-west-1.amazonaws.com/
Payload:
```
StackName: dev-kkm-adminApi-BaseNestedStackBaseNestedStackResourceAA4BD9B7-E3UA5TG3VQXV
LogicalResourceId: SSTMetadata
Action: DescribeStackResource
Version: 2010-05-15
```

```
<ErrorResponse xmlns="http://cloudformation.amazonaws.com/doc/2010-05-15/">
  <Error>
    <Type>Sender</Type>
    <Code>ValidationError</Code>
    <Message>Resource SSTMetadata does not exist for stack dev-kkm-adminApi-BaseNestedStackBaseNestedStackResourceAA4BD9B7-E3UA5TG3VQXV</Message>
  </Error>
  <RequestId>102a1ccf-4226-4753-9cbb-55c3e1c1bd88</RequestId>
</ErrorResponse>
```

I know that NestedStack is not implemented in SST, but without this modification, we could not further develop our API. Is it possible to make it work, or will it not work for us with nested stacks at all? It would be nice to at least ignore these parts so that we can work with other features in the console.

Thanks.

**Comment by romcok** [comment]
I solved this error by extending `cdk.NestedStack`. Now I don't see any errors in inspector, but it's still not working and it's stuck in "Syncing metadata".
```typescript

import * as path from "path";
import { Construct, IConstruct } from "constructs";
import * as cdk from "aws-cdk-lib";
import * as lambda from "aws-cdk-lib/aws-lambda";
import * as regionInfo from "aws-cdk-lib/region-info";
import { isConstruct } from "@serverless-stack/resources/dist/Construct";
import { App, Function, FunctionProps, Permissions } from "@serverless-stack/resources";
import { readFileSync } from 'fs'

export type SSTNestedStackProps = cdk.NestedStackProps;

export class SSTNestedStack extends cdk.NestedStack {
  public readonly stage: string;
  public readonly defaultFunctionProps: FunctionProps[];
  private readonly metadata: cdk.CfnResource;

  constructor(scope: Construct, id: string, props?: SSTNestedStackProps) {
    const root = scope.node.root as App;
    // const stackId = root.logicalPrefixedName(id);

    SSTNestedStack.checkForPropsIsConstruct(id, props);
    SSTNestedStack.checkForEnvInProps(id, props);

    super(scope, id, {
      ...props,
      /*env: {
        account: process.env.CDK_DEFAULT_ACCOUNT,
        region: root.region,
      },*/
    });

    this.stage = root.stage;
    this.defaultFunctionProps = root.defaultFunctionProps.map((dfp: any) =>
      typeof dfp === "function" ? dfp(this) : dfp
    );

    this.metadata = this.createMetadataResource();
  }

  public setDefaultFunctionProps(props: FunctionProps): void {
    const fns = this.getAllFunctions();
    if (fns.length > 0)
      throw new Error(
        "Default function props for the stack must be set before any functions have been added. Use 'addDefaultFunctionEnv' or 'addDefaultFunctionPermissions' instead to add more default properties."
      );
    this.defaultFunctionProps.push(props);
  }

  public addDefaultFunctionPermissions(permissions: Permissions) {
    this.defaultFunctionProps.push({
      permissions,
    });
  }

  public addDefaultFunctionEnv(environment: Record<string, string>) {
    this.defaultFunctionProps.push({
      environment,
    });
  }

  public addDefaultFunctionLayers(layers: lambda.ILayerVersion[]) {
    this.defaultFunctionProps.push({
      layers,
    });
  }

  public getAllFunctions() {
    return this.doGetAllFunctions(this);
  }

  private doGetAllFunctions(construct: IConstruct) {
    const results: Function[] = [];
    for (const child of construct.node.children) {
      if (child instanceof Function) results.push(child);
      results.push(...this.doGetAllFunctions(child));
    }
    return results;
  }

  public addOutputs(outputs: {
    [key: string]: string | cdk.CfnOutputProps;
  }): void {
    Object.keys(outputs).forEach((key) => {
      const value = outputs[key];
      if (value === undefined) {
        throw new Error(`The stack output "${key}" is undefined`);
      } else if (typeof value === "string") {
        new cdk.CfnOutput(this, key, { value });
      } else {
        new cdk.CfnOutput(this, key, value);
      }
    });
  }

  public addConstructsMetadata(metadata: any): void {
    this.metadata.addMetadata("sst:constructs", metadata);
  }

  private createMetadataResource(): cdk.CfnResource {
    // Add a placeholder resource to ensure stacks with just an imported construct
    // has at least 1 resource, so the deployment succeeds.
    // For example: users often create a stack and use it to import a VPC. The
    //              stack does not have any resources.
    //
    // Note that the "AWS::CDK::Metadata" resource does not exist in GovCloud
    // and a few other regions. In this case, we will use the "AWS::SSM::Parameter"
    // resource. It does not matter what resource type we use. All we are interested
    // in is the Metadata.
    const props = this.isCDKMetadataResourceSupported()
      ? {
        type: "AWS::CDK::Metadata",
      }
      : {
        type: "AWS::SSM::Parameter",
        properties: {
          Type: "String",
          Name: `/sst/${this.stackName}`,
          Value: "metadata-placeholder",
          Description: "Parameter added by SST for storing stack metadata",
        },
      };
    const res = new cdk.CfnResource(this, "SSTMetadata", props);
    
    // Add version metadata
    const packageJson = JSON.parse(readFileSync(
        path.join(__dirname, "..", "package.json")
    ).toString());
    res.addMetadata("sst:version", packageJson.version);

    return res;
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private static checkForPropsIsConstruct(id: string, props?: any) {
    // If a construct is passed in as stack props, let's detect it and throw a
    // friendlier error.
    if (props && isConstruct(props)) {
      throw new Error(
        `Expected an associative array as the stack props while initializing "${id}" stack. Received a construct instead.`
      );
    }
  }

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private static checkForEnvInProps(id: string, props?: any) {
    if (props && props.env) {
      let envS = "";

      try {
        envS = " (" + JSON.stringify(props.env) + ")";
      } catch (e) {
        // Ignore
      }

      throw new Error(
        `Do not set the "env" prop while initializing "${id}" stack${envS}. Use the "AWS_PROFILE" environment variable and "--region" CLI option instead.`
      );
    }
  }

  private isCDKMetadataResourceSupported(): boolean {
    const app = this.node.root as App;

    // CDK Metadata resource currently not supported in the region
    if (!regionInfo.RegionInfo.get(app.region).cdkMetadataResourceAvailable) {
      return false
    }

    // CDK Metadata resource used to not supported in the region
    // Note that b/c we cannot change the resource type of a given logical id,
    //           so if it used to not support, we will continue to mark it not
    //           supportd.
    if (['us-gov-east-1',
      'us-gov-west-1',
      'us-iso-east-1',
      'us-isob-east-1',
      'ap-northeast-3',
    ].includes(app.region)) {
      return false;
    }

    return true;
  }
}
```

**Comment by fwang** [comment]
Hey @romcok, oh wow.. thanks for exploring down this path.

As a side note, any reason to use nested stacks over having multiple stacks? You can spread the routes for an API across multiple stack, ie.
```js
// stackA
this.api = new sst.Api(...);

// stackB
stackA.api.addRoutes(...);

// stackC
stackA.api.addRoutes(...);
```

Multi-stack setup usually deploy much faster than a nested stack.

**Comment by romcok** [comment]
Hi, we are using the AppSync API. I looked at the possibilities of how to divide it into smaller units, but using NestedStack turned out to be effortless. We added SST later, so our API is not done via SST constructs and only uses Lambdas for debugging which is also great. 
Is this separation also possible with AppSync? I tried it for a while but without much success. It is possible that this is due to the design of our construct for AppSync, which was not built for stack separation. Now it takes us about 8 minutes to deploy this API with about 70 lambdas/resolvers. In the new version of CDK, we might not need those nested stacks anymore. I haven't checked it yet but splitting it into modular stacks cloud be better for us.
